### ISSUE TITLE
compositeImage is diffing scrollbars and failing tests

### ISSUE THREAD

**Issue by morganfeeney** [gemini-testing/gemini#747]
Whenever I use compositeImage I get failed tests due to the browser scrollbar being the only difference between reference and test images.

Example diff image attached:

![chrome_md diff](https://cloud.githubusercontent.com/assets/1900222/23697280/17f0901e-03df-11e7-8876-8a348222cffb.png)

**See configuration below:**

Output of `gemini --version`: 4.18.1
...

Contents of `.gemini.js` file:

```javascript
module.exports = {
    rootUrl: 'file:///Users/uidesign1/Documents',
    gridUrl: 'http://127.0.0.1:4444/wd/hub',
    compositeImage: true,
    calibrate: false,

    browsers: {
        chrome_lg: {
          windowSize: '1200x1024',
            desiredCapabilities: {
                browserName: 'chrome'
            }
        },
        chrome_md: {
          windowSize: '900x1024',
            desiredCapabilities: {
                browserName: 'chrome'
            }
        }
    }
};

```

Test source code:

```javascript
gemini.suite('box-test', (suite) => {
    suite.setUrl('/workspace/gemini-tests/index.html')
        .setCaptureElements('body')
        .capture('plain');
});

```

Command used to run the test:

```
gemini update
```
```
gemini test --reporter html --reporter flat
```

**Comment by sipayRT** [comment]
hi. You can hide scrollbar on your page with CSS:

```css
::-webkit-scrollbar
{
    width: 0 !important;
}
```

but it works only in chrome

**Comment by morganfeeney** [comment]
Hi, I have a bit of JS which does the trick:

```
gemini.suite('My account', (suite) => {
    suite.setUrl('/templates/my-account/account-summary.html')
    .before(function(actions, find) {
      actions.executeJS(function() {
        var style = document.createElement('style');
        var head = document.querySelector('head');
        var scrollBarStyle = document.createTextNode('::-webkit-scrollbar {display: none;}');
        head.appendChild(style);
        style.appendChild(scrollBarStyle);
      })
    })
    .setCaptureElements('body')
    .capture('plain');
});
```

I don't want to start changing the HTML documents by hiding the scrollbars as they are used for more than testing. Therefore a gemini/JS approach works best in this case.

I was under the impression that `calibrate` was a setting that would remove the browser UI, is this not the case and I have misunderstood the documentation?

**Comment by sipayRT** [comment]
> I was under the impression that calibrate was a setting that would remove the browser UI, is this not the case and I have misunderstood the documentation?

calibration needs only for refinement real screen sizes for some browsers (for example, ie8). You can switch it off for Chrome.

**Comment by morganfeeney** [comment]
Thanks, I have a couple of questions if you don't mind ;)

- Has this ever been an issue before? I expect this to be a recurring problem for lots of users.
- How can I make the following JS function available to Gemini, e.g.
```js
// Create function outside of suite scope
function scrollBarStyle() {
  var style = document.createElement('style');
  var head = document.querySelector('head');
  var scrollBarStyle = document.createTextNode('::-webkit-scrollbar {display: none;}');
  head.appendChild(style);
  style.appendChild(scrollBarStyle);  
}
// Then call function in gemini suite .before() function
gemini.suite('My account', (suite) => {
    suite.setUrl('/templates/my-account/account-summary.html')
    .before(function(actions, find) {
      actions.executeJS(function(window) {
       window.scrollBarStyle();
      })
    })
    .setCaptureElements('body')
    .capture('plain');
});
```
?

**Comment by sipayRT** [comment]
you can pass your function to the `executeJS`:

```js
function scrollBarStyle(window) {
    some actions here 
}

gemini.suite('My account', (suite) => {
    suite
        .setUrl('/templates/my-account/account-summary.html')
        .before((actions) => actions.executeJS(scrollBarStyle))
        .setCaptureElements('body')
        .capture('plain');
});
```

**Comment by morganfeeney** [comment]
Thanks for helping with the JS.

**Comment by morganfeeney** [comment]
I am still concerned about the scrollbars being an issue for others, I've now solved this problem but had to create a solution.
Is there any documentation explaining the issue, or should there be, or can it be solved in another way?

**Comment by sipayRT** [comment]
yeah, I  think we can add some good practice and examples to the [wiki](https://github.com/gemini-testing/gemini/wiki)

**Comment by maapteh** [comment]
On other browsers I can't hide the scrollbar. Since we do css regression tests also for browsers (like IE), Gemini will only work on Chrome (with above hack) when you do full page regression tests. Is there a solid solution?

**Comment by sipayRT** [comment]
> On other browsers I can't hide the scrollbar

you can also use `overflow-y: hidden;` instead

**Comment by ballwood** [comment]
@morganfeeney we had this problem recently. the issue surfaced on chrome on windows when emulating iPhone devices.

Managed to solve it by looking at the chrome driver source here and working out some of the other param names for the options for mobileEmulation that weren't visible on a first google: https://github.com/bayandin/chromedriver/blob/master/capabilities.cc#L152. 

Seems that anything with touch enabled will cause the bar to be shown. so we set mobile: false & touch: false. Seemed to work.

```
chrome_lg: {
    desiredCapabilities: {
        browserName: 'chrome'
        chromeOptions: {
            mobileEmulation: {
                width: 1024,
                height: 768,
                mobile: false,
                touch: false     
            }
         }
    }
}
```

Try that and see if it works. I wonder if the machine its running on is touch enabled?

**Comment by withinsight** [comment]
This is very much still an issue in Chrome/OSX, and should be removed for all Gemini users. I can't use Gemini currently because of this.

**Comment by morganfeeney** [comment]
@withinsight, I now use Backstop JS. By default this issue is non-existent.

**Comment by brondsem** [comment]
The CSS workarounds don't work for a touch device, since that type of scrollbar cannot be hidden: https://stackoverflow.com/questions/40733385/hiding-webkit-scrollbar-when-overflow-scrolling-touch-is-enabled

Setting `deviceMetrics` to `touch: false` seems like a way to deal with that, but I am using a named device like `deviceName: "Nexus 5"` because that includes a `pixelRatio` value.  And I haven't been able to successfully set an explicit `pixelRatio` in `deviceMetrics`

**Comment by j0tunn** [comment]
You can set `screenshotDelay` option in your config: https://github.com/gemini-testing/gemini/blob/master/doc/config.md
